File: 2025-09-14/wordFrequency.js
function getWords(paragraph) {
  
  return paragraph
    .split(/[ ,\.!]/)
    .filter(word => word !== "")
    .map(word => word.toLowerCase())
    .reduce((arr, word) => {
      const index = arr.findIndex(wordObject => wordObject.name === word);
    
      if (index === -1)
        arr.push({ name: word, count: 1 });
      else
        arr[index].count ++;

      return arr;
    }, [])
    .sort((a, b) => b.count - a.count)
    .slice(0, 3)
    .map(wordObject => wordObject.name);
}

console.log(getWords("Coding in Python is fun because coding Python allows for coding in Python easily while coding"));
console.log(getWords("I like coding. I like testing. I love debugging!"));
console.log(getWords("Debug, test, deploy. Debug, debug, test, deploy. Debug, test, test, deploy!"));

console.log(getWords("A b b c c c"));

File: 2025-09-14/word_frequency.py
def get_words(paragraph):
    import re
    from functools import reduce
    words = re.split(r"[ ,\.!]", paragraph)
    words = filter(lambda word: word != "", words)
    words = map(lambda word: word.lower(), words)
    words = reduce(build_array_of_dicts, words, [])
    words = sorted(words, key = lambda word_dict: word_dict['count'], reverse = True)[:3]
    words = map(lambda word_dict: word_dict['name'], words)

    return list(words)

def build_array_of_dicts(arr, word):
    
    index = next((i for i, word_dict in enumerate(arr) if word_dict['name'] == word), None)
    
    if index == None:
        arr.append({ 'name' : word, 'count' : 1 })
    else:
        arr[index]['count'] += 1

    return arr

print(get_words("Coding in Python is fun because coding Python allows for coding in Python easily while coding"))
print(get_words("I like coding. I like testing. I love debugging!"))
print(get_words("Debug, test, deploy. Debug, debug, test, deploy. Debug, test, test, deploy!"))

print(get_words("A b b c c c"))